File: RESTful i Nodejs - Express/RockPaperScissors.js
const express = require('express');
const app = express();

app.use(express.urlencoded({ extended: true }));
app.use(express.json()) // per analitzar les peticions HTTP que portin JSON al body

const llistaPartides = [];
const seguimentPartida = [];

let codis = [];
let partidesAcabades = [];
let moviments = ["pedra", "paper", "tisores"];

app.get('/', (req, res) => res.send('hola'));

app.post('/api/iniciarJoc/:codi', (req, res) => {
    let codi = parseInt(req.params.codi);

    if (codis.includes(codi)) res.send("Codi de partida ja existent.");
    else if (!codi || isNaN(codi)) res.send("Si us plau introdueix un codi vàlid.");
    else if (partidesAcabades.includes(codi)) res.send("Aquesta partida ja ha finalitzat");
    else {
        llistaPartides.push({ codi: codi, estatPartida: 'Iniciada.', jugadesJugador1: '', jugadesJugador2: '', wJug1: 0, wJug2: 0, guanyador: '' });
        seguimentPartida.push({ codi: codi, compt: 0, compt2: 0, tornJugador: 1, wJug1: 0, wJug2: 0, jugat: 1 })
        codis.push(codi);
        res.send(`Partida creada. Número de partida: ${codi}`);
    }
});

app.put('/api/jugarPartida/:codi', (req, res) => {
    let codi = parseInt(req.params.codi);
    let partida = llistaPartides.find(a => a.codi === codi);
    let seguiment = seguimentPartida.find(a => a.codi === codi);

    if (partidesAcabades.includes(codi)) {
        res.send(`Aquesta partida ja ha finalitzat. Guanyador: ${partida.guanyador}.`)
    } else {
        if (partida.estatPartida != "En joc") res.send("La partida amb codi " + codi + " no està en joc.");
        else if (codis.includes(codi) && (partida.estatPartida == "En joc") && (seguiment.compt == seguiment.compt2)) {
            let jugadesJugador1 = partida.jugadesJugador1.split(" ");
            let jugadesJugador2 = partida.jugadesJugador2.split(" ");
            seguiment.jugat = 1;
            partida.wJug1 = 0;
            partida.wJug2 = 0;

            for (let i = 0; i < jugadesJugador1.length - 1; i++) {
                if (jugadesJugador1[i] == jugadesJugador2[i]) {
                    partida.wJug1 = partida.wJug1;

                } else if ((jugadesJugador1[i] === 'pedra' && jugadesJugador2[i] === 'tisores') ||
                    (jugadesJugador1[i] === 'paper' && jugadesJugador2[i] === 'pedra') || (jugadesJugador1[i] === 'tisores' && jugadesJugador2[i] === 'paper')) {
                    partida.wJug1++;

                } else if (jugadesJugador2[i] === "") {
                    partida.wJug1++;
                } else {
                    partida.wJug2++;
                }
            }

            if (partida.wJug1 == 3) {
                partida.guanyador = "Jugador 1";
                partidesAcabades.push(codi);
                res.send("El jugador 1 guanya la partida. \n" + JSON.stringify(partida));
            } else if (partida.wJug2 == 3) {
                partida.guanyador = "Jugador 1";
                partidesAcabades.push(codi);
                res.send("El jugador 2 guanya la partida. \n" + JSON.stringify(partida));
            } else if (seguiment.compt >= 5 && seguiment.compt2 >= 5) {
                if (partida.wJug1 > partida.wJug2) {
                    partida.guanyador = "Jugador 1";
                    res.send("El jugador 1 guanya la partida. \n" + JSON.stringify(partida));
                } else if (partida.wJug1 > partida.wJug2) {
                    partida.guanyador = "Jugador 2";
                    res.send("El jugador 2 guanya la partida. \n" + JSON.stringify(partida));
                } else {
                    partida.guanyador = "Empat";
                    res.send("La partida finalitza en empat. \n" + JSON.stringify(partida));
                }
                partidesAcabades.push(codi);
            } else {
                res.send(partida);
            }

        } else if (seguiment.compt != seguiment.compt2) res.send("El jugador 2 encara no ha llençat.");
        else res.send("La partida amb codi " + codi + " no existeix");
    }
});

app.get('/api/consultarEstatPartida/:codi', (req, res) => {
    let codi = parseInt(req.params.codi);
    let partida = llistaPartides.find(a => a.codi === codi);

    if (codis.includes(codi) || partidesAcabades.includes(codi)) res.send(partida);
    else res.send(`Introdueix un codi vàlid. Codis de partides disponibles: ${codis.toString()}.`);
});



app.put('/api/moureJugador/:codi/:jugador/:jugada', (req, res) => { // /api/moureJugador
    let codi = parseInt(req.params.codi);
    let jugador = parseInt(req.params.jugador);
    let jugada = req.params.jugada;
    let seguiment = seguimentPartida.find(a => a.codi === codi);
    let partida = llistaPartides.find(a => a.codi === codi);

    if (!moviments.includes(jugada)) res.send("Jugada no vàlida. Introdueix un dels següents moviments: pedra, paper, tisores.");
    else if (!jugador || !(jugador > 0 && jugador <= 2)) res.send("Només pot haver-hi jugador 1 i jugador 2");
    else if (partidesAcabades.includes(codi)) res.send("Aquesta partida ja ha finalitzat.");
    else if (codis.includes(codi)) {
        partida.estatPartida = "En joc";

        if (seguiment.jugat == 0 && seguiment.tornJugador == 1) {
            res.send("Encara no s'ha jugat la partida.");
        }

        else if (jugador == 1) {
            if (seguiment.tornJugador != 1) {
                res.send('El jugador 2 encara no ha tirat.');
            } else {
                seguiment.jugat = 0;
                seguiment.compt++;
                partida.jugadesJugador1 += `${jugada} `;
                seguiment.tornJugador = 2;
                res.send(`Jugada ${seguiment.compt} executada.`);
            }
        }

        else if (jugador == 2) {
            if (seguiment.tornJugador != 2) {
                res.send('El jugador 1 encara no ha tirat.');
            } else {
                seguiment.compt2++;
                partida.jugadesJugador2 += `${jugada} `;
                seguiment.tornJugador = 1;
                console.log(partida);
                res.send(`Jugada ${seguiment.compt2} executada.`);
            }
        }
    }
    else res.send("Introdueix un codi vàlid.");

});

app.delete('/api/acabarJoc/:codi', (req, res) => {
    let codi = parseInt(req.params.codi);
    let seguiment = seguimentPartida.find(a => a.codi === codi);
    let partida = llistaPartides.find(a => a.codi === codi);

    if (codis.includes(codi)) {
        partidesAcabades.push(codi);
        codis.splice(codis.indexOf(codi), 1);
        partida.estatPartida = "Finalitzada";
        seguimentPartida.splice(seguimentPartida.indexOf(seguiment),1);
        res.send("Partida finalitzada. \n" + JSON.stringify(partida));
    } else if (partidesAcabades.includes(codi)) {
        codis.splice(codis.indexOf(codi), 1);
        partida.estatPartida = "Finalitzada";
        seguimentPartida.splice(seguimentPartida.indexOf(seguiment),1);
        res.send("Aquesta partida ja ha finalitzat.");
    } else res.send("La partida amb el codi " + codi + " no existeix");

});

app.listen(3001, () => console.log('Servidor iniciat.'));

File: RESTful JAX-RS/src/main/java/edu/fje/daw2/restfuljaxrs/RockPaperScissors.java
package edu.fje.daw2.restfuljaxrs;

import jakarta.ws.rs.*;
import jakarta.ws.rs.core.Context;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import jakarta.ws.rs.core.UriInfo;
import jakarta.servlet.*;
import jakarta.servlet.http.*;

import java.util.*;

@Path("daw2")
public class RockPaperScissors {

    @Context
    private UriInfo context;
    private static List<Partida> llistaPartides = new ArrayList<Partida>();
    private static List<Integer> codis = new ArrayList<Integer>();
    private static List<Integer> partidesAcabades = new ArrayList<Integer>();
    private static List<seguimentPartida> seguiment = new ArrayList<seguimentPartida>();

    public RockPaperScissors() {
        if (llistaPartides.size() == 0) {
            llistaPartides.add(new Partida(123,"Guillem", "Iniciada.","", 0,0,""));
            codis.add(123);
        }
    }

    @Path("/iniciarJoc/{codiPartida}")
    @POST
    @Produces(MediaType.TEXT_PLAIN)
    public String iniciarJoc(@PathParam("codiPartida") int codi) {
        if(codis.contains(codi)) {
            return "Ja existeix una partida amb aquest codi.";
        } else if (partidesAcabades.contains(codi)) {
            return "Aquesta partida ja ha finalitzat.";
        }
        codis.add(codi);
        Partida temp = new Partida(codi, "Iniciada","","",0,0,"");
        seguimentPartida temp2 = new seguimentPartida(codi, 0, 0,1,1);
        llistaPartides.add(temp);
        seguiment.add(temp2);
        System.out.println(llistaPartides.toString());
        return "Partida creada. Número de partida: "+codi+".";
    }

    @Path("/jugarPartida/{codiPartida}")
    @PUT
    @Produces(MediaType.TEXT_PLAIN)
    public String jugarPartida(@PathParam("codiPartida") int codi) {
        Partida partidaTemp = new Partida(codi, "","","",0,0,"");
        int pos = llistaPartides.indexOf(partidaTemp);
        seguimentPartida seguimentTemp = new seguimentPartida(codi,0 ,0,1,1);
        int pos2 = seguiment.indexOf(seguimentTemp);

        if (llistaPartides.get(pos).getEstatPartida().equals("Iniciada")) {
            return "La partida amb codi " + codi + " no està en joc.";
        } else if (codis.contains(codi) && llistaPartides.get(pos).getEstatPartida().equals("En joc") && seguiment.get(pos2).getCompt() == seguiment.get(pos2).getCompt2()) {
            seguiment.get(pos2).setJugat(1);

            String[] jugadesJugador1 = (llistaPartides.get(pos).getJugadesJugador1()).split(" ");
            String[] jugadesJugador2 = (llistaPartides.get(pos).getJugadesJugador2()).split(" ");
            System.out.println(Arrays.toString(jugadesJugador1));

            llistaPartides.get(pos).setwJug1(0);
            llistaPartides.get(pos).setwJug2(0);

            for (int i=0; i < jugadesJugador1.length; i++) {
                if (jugadesJugador1[i].equals(jugadesJugador2[i])) {
                    llistaPartides.get(pos).setwJug1(llistaPartides.get(pos).getwJug1());
                } else if (jugadesJugador1[i].equals("pedra") && jugadesJugador2[i].equals("tisores") || jugadesJugador1[i].equals("paper") && jugadesJugador2[i].equals("pedra") ||
                        jugadesJugador1[i].equals("tisores") && jugadesJugador2[i].equals("paper")) {
                    llistaPartides.get(pos).setwJug1(llistaPartides.get(pos).getwJug1()+1);
                } else if (jugadesJugador2[i].equals("")) {
                    llistaPartides.get(pos).setwJug1(llistaPartides.get(pos).getwJug1()+1);
                } else {
                    llistaPartides.get(pos).setwJug2(llistaPartides.get(pos).getwJug2()+1);
                }
            }

            if(llistaPartides.get(pos).getwJug1() == 3) {
                llistaPartides.get(pos).setGuanyador("Jugador 1");
                partidesAcabades.add(codi);
                return "Guanya jugador 1";
            } else if (llistaPartides.get(pos).getwJug2() == 3) {
                llistaPartides.get(pos).setGuanyador("Jugador 2");
                partidesAcabades.add(codi);
                return "Guanya jugador 2";
            } else {

            }

            return llistaPartides.get(pos).toString();
        } else if (seguiment.get(pos2).getCompt() != seguiment.get(pos2).getCompt2()) {
            return "El jugador 2 encara no ha llençat";
        } else if (partidesAcabades.contains(codi)) {
            return "Aquesta partida ja ha finalitzat. Guanyador: "+ llistaPartides.get(pos).getGuanyador();
        } else {
            return "La partida amb codi " + codi + " no existeix";
        }
    }

    @Path("/consultarEstatPartida/{codiPartida}")
    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String consultarEstatPartida(@PathParam("codiPartida") int codi) {
        if ((codis.contains(codi)) || partidesAcabades.contains(codi)) {
            Partida temp = new Partida(codi,"","","",0,0,"");
            int pos = llistaPartides.indexOf(temp);
            return llistaPartides.get(pos).toString();
        } else {
            return "La partida amb codi " + codi + " no existeix.";
        }

    }

    @Path("/moureJugador/{codiPartida}/{jugador}/{jugada}")
    @PUT
    @Produces(MediaType.TEXT_PLAIN)
    public String moureJugador(@PathParam("codiPartida") int codi, @PathParam("jugador") int jugador, @PathParam("jugada") String jugada){
        Partida temp = new Partida(codi,"","","",0,0,"");
        seguimentPartida temp2 = new seguimentPartida(codi,0 ,0,1,1);
        int pos = llistaPartides.indexOf(temp);
        int pos2 = seguiment.indexOf(temp2);
        llistaPartides.get(pos).setEstatPartida("En joc");

        if (!jugada.equals("pedra") && !jugada.equals("paper") && !jugada.equals("tisores")) return "Escull una jugada: pedra, paper o tisores.";
        else if (!(jugador > 0 && jugador <= 2)) return "Només pot haver-hi jugador 1 i jugador 2";
        else if (partidesAcabades.contains(codi)) return "Aquesta partida ja ha finalitzat.";
        else if (codis.contains(codi)) {
            if (seguiment.get(pos2).getCompt2() == 5 && seguiment.get(pos2).getCompt() == 5) {
                return "Partida finalitzada, acudeix a /acabarJoc per esbrinar el guanyador";
            } else if (seguiment.get(pos2).getJugat() == 0 && seguiment.get(pos2).getTornJugador() == 1) {
                return "Encara no s'ha jugat la partida";
            } else if(jugador == 1) {
                if (seguiment.get(pos2).getTornJugador() != 1) {
                    return "El jugador 2 encara no ha tirat.";
                } else {
                    seguiment.get(pos2).setJugat(0);
                    seguiment.get(pos2).setCompt(seguiment.get(pos2).getCompt() + 1);
                    llistaPartides.get(pos).setJugadesJugador1(llistaPartides.get(pos).getJugadesJugador1()+jugada+" ");
                    seguiment.get(pos2).setTornJugador(2);
                    if (seguiment.get(pos2).getCompt() >= 5) {
                        seguiment.get(pos2).setTornJugador(2);
                        seguiment.get(pos2).setCompt(5);
                        return "Jugada 5 executada. Partida finalitzada, acudeix a /acabarJoc per esbrinar el guanyador";
                    }
                    System.out.println(llistaPartides.get(pos).toString());
                    return "Jugada " +seguiment.get(pos2).getCompt()+" executada.";
                }
            }

            else {
                if (seguiment.get(pos2).getTornJugador() != 2) {
                    return "El jugador 1 encara no ha tirat.";
                } else {
                    seguiment.get(pos2).setCompt2(seguiment.get(pos2).getCompt2() + 1);
                    llistaPartides.get(pos).setJugadesJugador2(llistaPartides.get(pos).getJugadesJugador2()+jugada+" ");
                    seguiment.get(pos2).setTornJugador(1);
                    if (seguiment.get(pos2).getCompt2() >= 5) {
                        seguiment.get(pos2).setTornJugador(1);
                        seguiment.get(pos2).setCompt2(5);
                        return "Jugada 5 executada. Partida finalitzada, acudeix a /acabarJoc per esbrinar el guanyador";
                    }
                    System.out.println(llistaPartides.get(pos).toString());
                    return "Jugada " +seguiment.get(pos2).getCompt2()+" executada.";
                }
            }
        }
        return "Introdueix un codi vàlid";
    }

    @Path("/acabarJoc/{codiPartida}")
    @DELETE
    @Produces(MediaType.TEXT_PLAIN)
    public String acabarJoc(@PathParam("codiPartida") int codi) {
        Partida partidaTemp = new Partida(codi,"","","",0,0,"");
        int pos = llistaPartides.indexOf(partidaTemp);
        seguimentPartida seguimentTemp = new seguimentPartida(codi,0 ,0,1,1);

        if (codis.contains(codi)) {
            partidesAcabades.add(codi);
            codis.remove(codis.indexOf(codi));
            llistaPartides.get(pos).setEstatPartida("Finalitzada");
            return "Partida finalitzada. \n"+llistaPartides.get(pos).toString();
        }

        return llistaPartides.get(pos).toString();
    }
}

